tests/upgrades/scenario_workers.py

```diff
@@ -0,0 +1,58 @@
+import json
+import os
+
+import pytest
+from automation_tools.satellite6.hammer import set_hammer_config
+from fabric.api import env
+
+import robottelo
+from robottelo.config import settings
+
+
+_json_file = 'upgrade_workers.json'
+
+
+def save_worker_id(test_name, worker_id):
+    data = {}
+    if os.path.exists(_json_file):
+        with open(_json_file) as rjson:
+            data = json.load(rjson)
+    with open(_json_file, 'w') as wjson:
+        # Removing the parameter name from test name before save
+        test_name = test_name.split('[')[0] if '[' in test_name else test_name
+        data.update({test_name: worker_id})
+        wjson.seek(0)
+        json.dump(data, wjson)
+
+
+def get_worker_id(test_name):
+    with open(_json_file) as rjson:
+        data = json.load(rjson)
+        return data.get(test_name)
+
+
+@pytest.fixture(autouse=True)
+def save_pretest_worker_id(request, worker_id):
+    """Saves the worker id of pre_upgrade test in json"""
+    if request.node.get_closest_marker('pre_upgrade'):
+        save_worker_id(request.node.name, worker_id)
+
+
+@pytest.fixture(autouse=True)
+def set_posttest_worker_id(request):
+    """Retrieves worker_id of pre_upgrade and sets the same to run post_upgrade"""
```

Comment:
This docblock and function naming is misleading.

The function doesn't change the worker ID, it changes the target server hostname for the worker where this fixture is executing.

This fixture also breaks isolation of the satellite targets on their xdist workers, implicitly changing to a `balance` XDIST_BEHAVIOR, since multiple workers could end up setting the same `server.hostname` as a target.

The docblock needs to clearly state the limitation of this implementation, that XDIST_BEHAVIOR won't be observed, and tests running under xdist will not be isolated.

Reply:
Fixed!